First take at READMEs (#86)
canonical/dns-operators: dns-policy-operator/README.md

[![CharmHub Badge](https://charmhub.io/bind/badge.svg)](https://charmhub.io/dns-policy)
[![Publish to edge](https://github.com/canonical/dns-operators/actions/workflows/publish_charm.yaml/badge.svg)](https://github.com/canonical/dns-operators/actions/workflows/publish_charm.yaml)
[![Promote charm](https://github.com/canonical/dns-operators/actions/workflows/promote_charm.yaml/badge.svg)](https://github.com/canonical/dns-operators/actions/workflows/promote_charm.yaml)
[![Discourse Status](https://img.shields.io/discourse/status?server=https%3A%2F%2Fdiscourse.charmhub.io&style=flat&label=CharmHub%20Discourse)](https://discourse.charmhub.io)

# Dns Policy Operator

A [Juju](https://juju.is/) [charm](https://juju.is/docs/olm/charmed-operators)
deploying and managing a DNS server on bare metal.

This charm simplifies configuration of a DNS server by providing a single point
of configuration for all the requirers using the same DNS server. It can be
deployed on virtual machines or bare metal.

As such, the charm makes it easy to manage and propagate DNS configuration, while

> giving the freedom to deploy on the substrate of their choice.

Reviewer: I noticed that this description is identical to what's in the bind-operator README. I'm wondering whether the description here should emphasize the policy part of this charm. 

So, instead of `This charm simplifies configuration of a DNS server by providing a single point
of configuration for all the requirers using the same DNS server.`, the sentence here could say something like, `This charm simplifies configuration of a DNS server by providing record requests to all requirers using the same DNS server.` (If that's technically incorrect, please let me know, and update accordingly!) 

Same thing with `the charm makes it easy to manage and propagate DNS configuration` -- maybe `the charm makes it easy to manage and propagate DNS record requests`, assuming that's the correct interpretation of this charm

Author: I had completely forgotten to rewrite that. It is now done

For DevOps or SRE teams this charm will make operating any charm requiring dynamic DNS
configuration simple and straightforward through Juju's clean interface.

## Get started

> The charm can be deployed to any machine model by pulling it from [charmhub](https://charmhub.io/bind):

Reviewer: ```suggestion
The charm can be deployed to any machine model by pulling it from [Charmhub](https://charmhub.io/bind):
```

```
juju deploy dns-policy --channel=latest/edge
```

You then need to integrate it with a charm supporting the `dns_record` interface for it to work as it is a subordinate charm.  

> 
> ### Basic operations
> 
> Before beeing able to review incoming dns record requests, you will need to create a reviewer account.  
> To do that, use the `create-reviewer` command and then log in using it's credentials to the dns-policy interface.  

Reviewer: ```suggestion
Before being able to review incoming DNS record requests, you will need to create a reviewer account.  
To do that, use the `create-reviewer` command, and then log in using its credentials to the dns-policy interface.  

### Basic operations

```
Is it possible to show an example of what the `create-reviewer` command looks like in action?

Author: I added some basic operations descriptions

> It can integrate with any requirer charm implementing the [dns_record interface](https://canonical.github.io/charm-relation-interfaces/interfaces/dns_record/v0/).

Reviewer: ```suggestion
The charm can integrate with any requirer charm implementing the [dns_record interface](https://canonical.github.io/charm-relation-interfaces/interfaces/dns_record/v0/).
```

> This charm will be give your reviewers the ability to manage all the DNS record requests from those requirer. Only one provider can be integrated though.

Reviewer: ```suggestion
This charm will be give your reviewers the ability to manage all the DNS record requests from those requirer. Only one provider can be integrated.
```
Nitpick

## Learn more
* [Read more](https://charmhub.io/dns-policy/docs)

> * [Official webpage](https://charmhub.io/dns-policy)

Reviewer: Same comment about potentially removing the "Official website" line

## Project and community
* [Issues](https://github.com/canonical/dns-operators/issues)

> * [Contributing](https://charmhub.io/bind/docs/how-to-contribute)

Reviewer: ```suggestion
* [Contributing](https://github.com/canonical/dns-operators/blob/main/CONTRIBUTING.md)
```

[![CharmHub Badge](https://charmhub.io/bind/badge.svg)](https://charmhub.io/dns-policy)
[![Publish to edge](https://github.com/canonical/dns-operators/actions/workflows/publish_charm.yaml/badge.svg)](https://github.com/canonical/dns-operators/actions/workflows/publish_charm.yaml)
[![Promote charm](https://github.com/canonical/dns-operators/actions/workflows/promote_charm.yaml/badge.svg)](https://github.com/canonical/dns-operators/actions/workflows/promote_charm.yaml)
[![Discourse Status](https://img.shields.io/discourse/status?server=https%3A%2F%2Fdiscourse.charmhub.io&style=flat&label=CharmHub%20Discourse)](https://discourse.charmhub.io)

> # Dns Policy Operator

Reviewer: ```suggestion
# DNS Policy Operator
```

A [subordinate](https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/charm/index.html#subordinate) [Juju](https://juju.is/) [charm](https://juju.is/docs/olm/charmed-operators)
enabling a policy layer on top of a DNS charm like [Bind](https://github.com/canonical/dns-operators/tree/main/bind-operator).

After deploying the Bind operator, you may want to restrict the possibility of any charm integrating with it and freely modifying the DNS records it serves. You may want to add formal human approval for each incoming DNS resource record request. That's when you want to use the DNS Policy charm.

> By deploying this charm and integrating it with the Bind operator, you create a policy layer that can be operated using a web interface and/or an API, all provided by the Django workload operated by it.

Reviewer: ```suggestion
By deploying this charm and integrating it with the Bind operator, you create a policy layer that can be operated using a web interface or an API, all provided by the Django workload operated by the charm.
```
Trying to clarify what "it" is -- please update if I misunderstood

## Get started

The charm can be deployed to any machine model by pulling it from [Charmhub](https://charmhub.io/bind):
```
juju deploy dns-policy --channel=latest/edge
```

You then need to integrate it with a charm supporting the `dns_record` interface for it to work as it is a subordinate charm.  

> Before beeing able to review incoming dns record requests, you will need to create a reviewer account.  

Reviewer: ```suggestion
Before being able to review incoming DNS record requests, you will need to create a reviewer account.  
```

> To do that, use the `create-reviewer` command and then log in using it's credentials to the dns-policy interface.  

Reviewer: ```suggestion
To do that, use the `create-reviewer` command and then log in using its credentials to the dns-policy interface.  
```

The charm can integrate with any requirer charm implementing the [dns_record interface](https://canonical.github.io/charm-relation-interfaces/interfaces/dns_record/v0/).
This charm will be give your reviewers the ability to manage all the DNS record requests from those requirer. Only one provider can be integrated.

### Basic operations

#### Create a reviewer

> Once the charm deployed and running, you can create a reviewer with the following command:  

Reviewer: ```suggestion
Once the charm is deployed and running, you can create a reviewer with the following command:  
```

> `juju run <dns-policy-unit> create-reviewer username=<reviewer-username> email=<reviewer-email>`  

Reviewer: I suggest to put this into a code block (with three backticks -- I would use the "suggestion" feature on GitHub but it breaks when trying to suggest code blocks)

The action will generate a password and display it once the user has been created.

Example:
```
$ juju run dns-policy/0 create-reviewer username=reviewer email=reviewer@example.com
Running operation 3 with 1 task
  - task 4 on unit-dns-policy-0

Waiting for task 4...
result: |
  User reviewer created successfully
  Generated password: znaLEEYA6s9sfFht

> ```

Reviewer: The [style guide discourages using prompt marks in code blocks](https://docs.ubuntu.com/styleguide/en/#code-examples-in-documentation). I suggest that you rephrase to say something like below:

-------------------------

For example, by running the `create-reviewer` action with `username=reviewer`, the terminal will respond with something similar to the following:
```
Running operation 3 with 1 task
  - task 4 on unit-dns-policy-0
Waiting for task 4...
result: |
  User reviewer created successfully
  Generated password: znaLEEYA6s9sfFht
```

Reviewer: Your suggestion works for me!

#### Restrict incoming requests

The workload of the DNS policy charm is a Django application. As such, it can be restricted on incoming requests
using the `allowed-hosts` configuration.  

> Example: `juju config dns-policy allowed-hosts="1.2.3.4"`  

Reviewer: I would put this example in a dedicated code block

This will restrict incoming requests to only `1.2.3.4`.

> More informations about allowed hosts in the [Django documentation](https://docs.djangoproject.com/en/5.1/ref/settings/#allowed-hosts).

Reviewer: ```suggestion
See more information about allowed hosts in the [Django documentation](https://docs.djangoproject.com/en/5.1/ref/settings/#allowed-hosts).
```

#### Set logging

The workload snap and underlying Django application can be set to debug mode by setting the debug configuration.  

> Example: `juju config dns-policy debug=true`  

Reviewer: I would put this example in a dedicated code block

> More informations about debug mode in the [Django documentation](https://docs.djangoproject.com/en/5.1/ref/settings/#debug).

Reviewer: ```suggestion
See more information about debug mode in the [Django documentation](https://docs.djangoproject.com/en/5.1/ref/settings/#debug).
```
